### PR TITLE
Added exttrigger capability as type 7

### PR DESCRIPTION
**Pull request by pdonahue-ventana**
Enables cross triggering between debug and other blocks such as hardware performance monitors.
Skipped over type 6 assuming #538 will be approved.

### CODE REVIEW

File: trigger.tex
Comment (timsifive) [inline trigger.tex:72]: I'd omit "over an external debug trigger output". Is it important that this is a pulse? That sounds like an implementation detail, but if there's already something elsewhere that says hpmcounters can have inputs that are pulsed then it would make sense to use the same definition here.

File: xml/hwbp_registers.xml
Comment (timsifive) [inline xml/hwbp_registers.xml:577]: Would it be better to make this a 16-bit field, allowing the trigger to fire when any of the selected inputs sends a signal?

File: xml/hwbp_registers.xml
Comment (timsifive) [inline xml/hwbp_registers.xml:559]: I'd use "signals" instead of "asserts," and remove the parenthetical.

**Review by timsifive — CHANGES_REQUESTED** [review]
Looks like a reasonable way to implement this feature. Comments below.

Please don't submit riscv-debug-draft.pdf in your pull requests. This leads to annoying conflicts, and generally git isn't great at maintaining lots of binary history.

I manually rebuild the PDF once a month, just before the meeting. That seems to be enough to give people who don't want to build a PDF something to use.

File: trigger.tex
Comment (pdonahue-ventana) [reply trigger.tex]: I updated this.

File: xml/hwbp_registers.xml
Comment (pdonahue-ventana) [reply xml/hwbp_registers.xml]: Sounds good.  The downside is that there are fewer bits for other purposes but I don't think there's a need for a large number of additional bits.

File: xml/hwbp_registers.xml
Comment (pdonahue-ventana) [reply xml/hwbp_registers.xml]: I updated this and specified that it fires if "any" selected input signals to be explicit that this is an OR function.

**Review by timsifive — APPROVED** [review]
Looks good. Thanks! I'll merge this Friday unless there are more comments. @ernie-sifive, do you have anything to add?

File: xml/hwbp_registers.xml
Comment (ernie-sifive) [inline xml/hwbp_registers.xml:564]: I wonder if we should use a name other than "external debug trigger" to avoid confusing these with the external triggers in the debug module referred to in dmcs2 and used for halt and resume groups.  Maybe something like intra-hart logic trigger?

Comment (pdonahue-ventana) [reply xml/hwbp_registers.xml:564]: Sure.  Intra-hart logic trigger is fine with me.

There's the description but also the register name.  This register name should be similar to the description and should probably avoid using the same name as the dmcs2 exttrigger field (which it currently conflicts with).  What should this register be called?
"intrahartlogictrigger" is too long.
"ihlt" is too short.
"ihltrigger"?
"iharttrigger"?  (The unicode "I❤️trigger" is probably out. 😄)
"intrahart"?

Or maybe there's a description other than "intra-hart logic trigger" that lends itself to a more natural name.

Comment (timsifive) [reply xml/hwbp_registers.xml:564]: This is a tricky one. Intra-hart seems limiting. These triggers could be connected to something outside the RISC-V core altogether, and that should be fine.

They really are external triggers, going to a hart. The external triggers in dmcs2 are also external triggers, going to a Debug Module. Should we rename both? And then the direction matters, too.

dmcs2.exttrigger refers to input/output triggers to/from the DM.
exttrigger.select refers to input triggers to the hart.

I'm not arriving at a great name here, but just sharing some thoughts.

**Comment by timsifive** [comment]
@pdonahue-ventana, what are your thoughts on the naming issue?

**Comment by ernie-sifive** [comment]
How about "Hart Logic Trigger" or simply "Logic Trigger", then explicitly state that the logic signal can come from a source within the hart or from outside the hart.  Problem is this suggests that they have to be logic signals and not abstractions that might be from a message-passing network or something else.

Another possibility is to rename existing external triggers to "DM External Trigger" or "Debug External Trigger" and then call these new triggers "Hart External Trigger".

**Comment by timsifive** [comment]
What if we just call them "dm trigger" and "hart trigger?" I think external doesn't really add anything, and if somebody wants to wire these up to sources that are not external that's probably also fine.

**Comment by pdonahue-ventana** [comment]
I think Ernie's "DM External Trigger" sounds good since it's for things external to the DM.  I'd similarly call the new one "TM External Trigger" since it's for things that are external to the TM, regardless of whether they're coming from something associated with this hart (like HPM or trace) or from outside the hart.  I prefer this over "hart trigger" because I consider the internal TM triggers (load/store, pc, single step, etc.) to be hart triggers.

Thoughts?

**Comment by ernie-sifive** [comment]
@pdonahue-ventana I agree - we need terminology that indicates these entities are signal inputs rather than a block containing registers and comparators that can cause breakpoints (which is what I would call a "trigger").

**Comment by timsifive** [comment]
So "DM external trigger" and "TM external trigger" then?

**Comment by pdonahue-ventana** [comment]
Sounds good to me.  Does this mean that this register should be tmexttrigger and the dmcs2 field should change from exttrigger to dmexttrigger (plus some minor description changes)?

**Comment by pdonahue-ventana** [comment]
That change just now was to resolve merge conflicts.  I won't change names until I hear back.

**Review by ernie-sifive — APPROVED** [review]
Thanks, that looks good to me.

**Review by timsifive — CHANGES_REQUESTED** [review]
Just 1 comment.

File: xml/hwbp_registers.xml
Comment (pdonahue-ventana) [reply xml/hwbp_registers.xml:52]: Good catch.  I updated the name of chapter 5.  An acronym section would definitely be nice but I'll leave that to another ambitious person.

**Review by timsifive — APPROVED** [review]
Looks good. I'll merge this tomorrow unless there are more comments.

**Pull request merged**
